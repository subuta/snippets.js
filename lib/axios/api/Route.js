import { build, format, snippets as s, EOL } from 'bld.js'
import pluralize from 'pluralize'
import _ from 'lodash'

import getControllerActions from '../../config/getControllerActions'
import UpdateAction from './actions/Update'
import ShowAction from './actions/Show'
import DestroyAction from './actions/Destroy'
import IndexAction from './actions/All'
import CreateAction from './actions/Create'

export default (RouteConfig, ModelConfig) => {
  const {
    schema
  } = ModelConfig

  const {
    tableName
  } = schema

  const renderAction = (action) => {
    switch (action) {
      case 'index':
        return IndexAction(ModelConfig) + s.EOL
      case 'show':
        return ShowAction(ModelConfig) + s.EOL
      case 'create':
        return CreateAction(ModelConfig) + s.EOL
      case 'update':
        return UpdateAction(ModelConfig) + s.EOL
      case 'destroy':
        return DestroyAction(ModelConfig) + s.EOL
    }
  }

  const keys = getControllerActions(RouteConfig)
  const exports = _.zipObject(keys, keys)

  const actions = _.map(keys, renderAction)

  const model = pluralize.singular(tableName)
  const modelSchema = `${model}`
  const modelListSchema = `${modelSchema}List`

  const imports = s.import([
    ['lodash', '_'],
    ['src/utils/request', 'request'],
    ['normalizr', null, [
      ['normalize']
    ]],
    ['src/utils/schema', null, [
      [modelSchema],
      [modelListSchema],
    ]]
  ])

  return build`
    ${imports}
    
    ${actions}
    
    /* mat Custom action [start] */
    /* mat Custom action [end] */
    
    let actions = ${s.raw(exports)}
    
    /* mat Custom exports [start] */
    /* mat Custom exports [end] */
    
    ${s.export('actions')}
  `
}
