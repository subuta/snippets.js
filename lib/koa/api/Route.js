import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import IndexAction from './actions/All'
import ShowAction from './actions/Show'
import CreateAction from './actions/Create'
import UpdateAction from './actions/Update'
import DestroyAction from './actions/Destroy'

import getControllerActions from '../lib/_utils/getControllerActions'

export default (props) => {
  let {
    model,
    config
  } = props

  const {
    imports = []
  } = config

  // Ensure naming convention.
  model = pluralize.singular(model)
  const models = pluralize(model)

  const Import = s.import([
    ...imports,
    ['koa-router', 'Router'],
    ['lodash', '_'],
  ])

  const renderAction = (action) => {
    switch (action) {
      case 'index':
        return IndexAction(props)
      case 'show':
        return ShowAction(props)
      case 'create':
        return CreateAction(props)
      case 'update':
        return UpdateAction(props)
      case 'destroy':
        return DestroyAction(props)
    }
  }

  const actions = _.map(getControllerActions(config), renderAction)

  const prefix = config.prefix || `/${models}`

  return build`
    ${Import}
    
    const ${model} = new Router({
      prefix: ${s.stringify(prefix)}
    })
    
    ${actions}
    
    /* mat Custom actions [start] */
    /* mat Custom actions [end] */
    
    ${s.export(build`
      {
        routes: () => _.cloneDeep(${model}.routes()),
        register: (routers) => {
          /* mat Register [start] */
          /* mat Register [end] */
        }
      }
    `)}
  `
}
