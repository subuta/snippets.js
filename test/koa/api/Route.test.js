/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'

import Route from 'lib/koa/api/Route'

describe('knex/api/Route', () => {
  it('should create koa Route', () => {
    expect(format(Route({
      model: 'user',
      config: {}
    }))).toMatchSnapshot()
  })

  it('should create koa Route with custom prefix', () => {
    expect(format(Route({
      model: 'user',
      config: {
        prefix: '/hoge/fuga/users'
      }
    }))).toMatchSnapshot()
  })
})
