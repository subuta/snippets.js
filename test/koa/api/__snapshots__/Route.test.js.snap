// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`knex/api/Route should create koa Route 1`] = `
"import Router from 'koa-router'
import _ from 'lodash'

const user = new Router({
  prefix: '/users'
})

user.get('/', async (ctx) => {
  const {User} = ctx.state.models
  let params = {}

  /* mat Before index [start] */
  /* mat Before index [end] */

  ctx.body = await User.query()
    .eager('')
    .where(params)
})

user.get('/:id', async (ctx) => {
  const {User} = ctx.state.models
  let params = {}

  /* mat Before show [start] */
  /* mat Before show [end] */

  ctx.body = await User.query()
    .eager('')
    .findFirst({...params, id: ctx.params.id})
})

user.post('/', async (ctx) => {
  const {User} = ctx.state.models
  const {user} = ctx.request.body

  let params = {}

  /* mat Before create [start] */
  /* mat Before create [end] */

  let response = await User.query()
    .insert({
      ...user,
      ...params
    })
    .eager('')

  /* mat After create [start] */
  /* mat After create [end] */

  ctx.body = response
})

user.put('/:id', async (ctx) => {
  const {User} = ctx.state.models
  const {user} = ctx.request.body
  const {sub} = ctx.state.user

  // update specified user.
  const params = {}

  /* mat Before update [start] */
  /* mat Before update [end] */

  ctx.body = await User.query()
    .patchAndFetchById(ctx.params.id, {
      ...user,
      ...params
    })
    .eager('')
})

user.delete('/:id', async (ctx) => {
  const {User} = ctx.state.models
  await User.query()
    .delete()
    .where({id: ctx.params.id})
  ctx.body = null
})

/* mat Custom actions [start] */
/* mat Custom actions [end] */

export default {
  routes: () => _.cloneDeep(user.routes()),
  register: (routers) => {
    /* mat Register [start] */
    /* mat Register [end] */
  }
}
"
`;

exports[`knex/api/Route should create koa Route with custom prefix 1`] = `
"import Router from 'koa-router'
import _ from 'lodash'

const user = new Router({
  prefix: '/hoge/fuga/users'
})

user.get('/', async (ctx) => {
  const {User} = ctx.state.models
  let params = {}

  /* mat Before index [start] */
  /* mat Before index [end] */

  ctx.body = await User.query()
    .eager('')
    .where(params)
})

user.get('/:id', async (ctx) => {
  const {User} = ctx.state.models
  let params = {}

  /* mat Before show [start] */
  /* mat Before show [end] */

  ctx.body = await User.query()
    .eager('')
    .findFirst({...params, id: ctx.params.id})
})

user.post('/', async (ctx) => {
  const {User} = ctx.state.models
  const {user} = ctx.request.body

  let params = {}

  /* mat Before create [start] */
  /* mat Before create [end] */

  let response = await User.query()
    .insert({
      ...user,
      ...params
    })
    .eager('')

  /* mat After create [start] */
  /* mat After create [end] */

  ctx.body = response
})

user.put('/:id', async (ctx) => {
  const {User} = ctx.state.models
  const {user} = ctx.request.body
  const {sub} = ctx.state.user

  // update specified user.
  const params = {}

  /* mat Before update [start] */
  /* mat Before update [end] */

  ctx.body = await User.query()
    .patchAndFetchById(ctx.params.id, {
      ...user,
      ...params
    })
    .eager('')
})

user.delete('/:id', async (ctx) => {
  const {User} = ctx.state.models
  await User.query()
    .delete()
    .where({id: ctx.params.id})
  ctx.body = null
})

/* mat Custom actions [start] */
/* mat Custom actions [end] */

export default {
  routes: () => _.cloneDeep(user.routes()),
  register: (routers) => {
    /* mat Register [start] */
    /* mat Register [end] */
  }
}
"
`;
